File: src/calculus_area_between_curves_medium.py
"""
calculus - area_between_curves (medium)
Generated: 2026-02-22T05:16:14.607216
"""

from problem_utils import *

def generate():
    problem_type = choice(['given_bounds', 'parabolas_intersect', 'sqrt_linear', 'trig_const'])

    if problem_type == 'given_bounds':
        # Two polynomials on given interval, coefficients varied widely
        # f(x) = a1*x^2 + b1*x + c1  (upper)
        # g(x) = a2*x^2 + b2*x + c2  (lower, c2 < c1 to ensure f > g at interval start)
        a1 = randint(1, 4)
        b1 = randint(-4, 4)
        c1 = randint(1, 8)

        a2 = randint(1, 3)
        b2 = randint(-4, 4)
        c2 = c1 - randint(2, 8)   # ensures f above g on short intervals

        f = a1 * x**2 + b1 * x + c1
        g = a2 * x**2 + b2 * x + c2

        x_left = randint(0, 3)
        x_right = x_left + randint(1, 3)

        integrand = simplify(f - g)
        ans = integrate(integrand, (x, x_left, x_right))
        ans = simplify(ans)

        question = (f"Find the area between $f(x) = {latex(f)}$ and $g(x) = {latex(g)}$ "
                    f"from $x = {x_left}$ to $x = {x_right}$.")

        solution = steps(
            f"Area $= \\int_{{{x_left}}}^{{{x_right}}} [f(x) - g(x)] \\, dx$",
            f"$f(x) - g(x) = {latex(integrand)}$",
            f"$= \\left[{latex(integrate(integrand, x))}\\right]_{{{x_left}}}^{{{x_right}}}$",
            f"$= {latex(ans)}$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1300, 1450),
            topic="calculus/area_between_curves",
            solution=solution,
            calculator="scientific"
        )

    elif problem_type == 'parabolas_intersect':
        # f(x) = -a*x² + p,  g(x) = b*x² - q  — find intersections then integrate
        a_coeff = randint(1, 4)
        b_coeff = randint(1, 4)
        # Choose p, q so intersection at x = ±r for integer r
        r = randint(1, 4)
        total = (a_coeff + b_coeff) * r**2
        p = total // 2 + randint(0, 3)
        q = total - p

        f = -a_coeff * x**2 + p
        g = b_coeff * x**2 - q

        intersections = solve(Eq(f, g), x)
        intersections_sorted = sorted([pt for pt in intersections], key=lambda v: float(v.evalf()))
        x_left, x_right = intersections_sorted[0], intersections_sorted[-1]

        integrand = simplify(f - g)
        ans = simplify(integrate(integrand, (x, x_left, x_right)))

        question = f"Find the area between $f(x) = {latex(f)}$ and $g(x) = {latex(g)}$."

        solution = steps(
            f"Find intersections: ${latex(f)} = {latex(g)}$",
            f"${latex(simplify(f - g))} = 0$",
            f"$x = {latex(x_left)},\\; x = {latex(x_right)}$",
            f"Area $= \\int_{{{latex(x_left)}}}^{{{latex(x_right)}}} [{latex(integrand)}] \\, dx$",
            f"$= {latex(ans)}$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1400, 1550),
            topic="calculus/area_between_curves",
            solution=solution,
            calculator="scientific"
        )

    elif problem_type == 'sqrt_linear':
        # f(x) = c*√x  and  g(x) = m*x  on [0, c²/m²] (their intersection)
        c = randint(1, 5)
        m = randint(1, 4)

        # Intersect: c√x = mx → c²x = m²x² → x = c²/m²
        x_right_val = Rational(c**2, m**2)
        x_left_val = 0

        f = c * sqrt(x)
        g = m * x

        integrand = simplify(f - g)
        ans = simplify(integrate(integrand, (x, x_left_val, x_right_val)))

        question = f"Find the area enclosed between $f(x) = {c}\\sqrt{{x}}$ and $g(x) = {m}x$."

        solution = steps(
            f"Find intersections: ${c}\\sqrt{{x}} = {m}x$",
            f"Square: ${c**2}x = {m**2}x^2$ → $x = 0$ or $x = {latex(x_right_val)}$",
            f"For $0 \\leq x \\leq {latex(x_right_val)}$: $f(x) \\geq g(x)$",
            f"Area $= \\int_{{0}}^{{{latex(x_right_val)}}} [{c}\\sqrt{{x}} - {m}x] \\, dx$",
            f"$= {latex(ans)}$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1450, 1600),
            topic="calculus/area_between_curves",
            solution=solution,
            calculator="scientific"
        )

    else:  # trig_const
        # Area between A*sin(bx) and constant k on [0, π/b]
        A = randint(1, 5)
        b = randint(1, 4)
        k = randint(0, A - 1) if A > 1 else 0

        f = A * sin(b * x)
        g = k
        lower = 0
        upper = pi / b

        integrand = f - g
        ans = simplify(integrate(integrand, (x, lower, upper)))

        question = (f"Find the area between $f(x) = {A}\\sin({b}x)$ and $y = {k}$ "
                    f"from $x = 0$ to $x = \\dfrac{{\\pi}}{{{b}}}$.")

        solution = steps(
            f"On $[0, \\pi/{b}]$, $\\sin({b}x) \\geq 0$ so $f(x) \\geq {k}$ (check endpoints)",
            f"Area $= \\int_{{0}}^{{\\pi/{b}}} [{A}\\sin({b}x) - {k}] \\, dx$",
            f"$= \\left[-\\frac{{{A}}}{{{b}}}\\cos({b}x) - {k}x\\right]_{{0}}^{{\\pi/{b}}}$",
            f"$= {latex(ans)}$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1350, 1550),
            topic="calculus/area_between_curves",
            solution=solution,
            calculator="scientific"
        )

emit(generate())

File: src/calculus_area_between_curves_medium.jl
# calculus - area_between_curves (medium)
# Generated: 2026-03-08T20:48:03.877696

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("calculus/area_between_curves")
    
    # Choose problem type
    prob_type = choice([:linear_quadratic, :two_quadratics, :quadratic_cubic, :polynomial_shifted])
    
    if prob_type == :linear_quadratic
        # Linear and quadratic - clean intersection points
        r1 = randint(-8, 8)
        r2 = r1 + randint(2, 6)
        
        # Top function: parabola opening down through both points
        a_top = -randint(1, 3)
        # Use vertex form shifted to be between the roots
        h = (r1 + r2) / 2
        k = randint(3, 12)
        f_top = expand(a_top * (x - h)^2 + k)
        
        # Bottom function: line through both points
        y1 = substitute(f_top, x => r1)
        y2 = substitute(f_top, x => r2)
        m = (y2 - y1) // (r2 - r1)
        b = y1 - m * r1
        f_bottom = m * x + b
        
        # Calculate area
        integrand = expand(f_top - f_bottom)
        antideriv = simplify(integrate(integrand, x))
        area_val = substitute(antideriv, x => r2) - substitute(antideriv, x => r1)
        area_val = abs(area_val)
        
        problem(
            question="Find the area between the curves \$f(x) = $(tex(f_top))\$ and \$g(x) = $(tex(f_bottom))\$ from \$x = $(r1)\$ to \$x = $(r2)\$.",
            answer=area_val,
            difficulty=(1200, 1400),
            solution=steps(
                sol("Top function", f_top),
                sol("Bottom function", f_bottom),
                sol("Set up integral", "A = \\int_{$(r1)}^{$(r2)} [$(tex(f_top)) - ($(tex(f_bottom)))] dx"),
                sol("Integrand", integrand),
                "Integrate and evaluate at bounds",
                sol("Area", area_val)
            ),
            time=180
        )
        
    elseif prob_type == :two_quadratics
        # Two parabolas with clean intersection
        r1 = randint(-6, 0)
        r2 = r1 + randint(4, 8)
        
        # Top: opens down
        a1 = -randint(1, 2)
        q1 = rand_factorable(x; a=(a1, a1), roots=(r1, r2))
        
        # Bottom: opens up, vertex below
        a2 = randint(1, 3)
        shift = -randint(5, 15)
        q2_base = rand_factorable(x; a=(a2, a2), roots=(r1, r2))
        q2 = expand(q2_base.expr + shift)
        
        integrand = expand(q1.expr - q2)
        antideriv = integrate(integrand, x)
        area_val = substitute(antideriv, x => r2) - substitute(antideriv, x => r1)
        area_val = abs(simplify(area_val))
        
        problem(
            question="Find the area of the region enclosed by \$y = $(tex(q1.expr))\$ and \$y = $(tex(q2))\$.",
            answer=area_val,
            difficulty=(1400, 1600),
            solution=steps(
                "Find intersection points by setting the functions equal",
                sol("Intersection points", "x = $(r1), x = $(r2)"),
                sol("Top function", q1.expr),
                sol("Bottom function", q2),
                sol("Set up integral", "A = \\int_{$(r1)}^{$(r2)} [($(tex(q1.expr))) - ($(tex(q2)))] dx"),
                "Integrate and evaluate",
                sol("Area", area_val)
            ),
            time=240
        )
        
    elseif prob_type == :quadratic_cubic
        # Quadratic and cubic with two intersection points
        r1 = randint(-5, 0)
        r2 = r1 + randint(3, 6)
        
        # Cubic through both points
        a_cub = choice([-1, 1])
        # Simple cubic: a(x - r1)(x - r2)(x - c) where c is between r1 and r2
        c = r1 + randint(1, r2 - r1 - 1)
        f_cubic = expand(a_cub * (x - r1) * (x - r2) * (x - c))
        
        # Quadratic through both points
        y1 = substitute(f_cubic, x => r1)
        y2 = substitute(f_cubic, x => r2)
        
        # Use vertex form for quadratic
        h = (r1 + r2) / 2
        a_quad = -a_cub * randint(2, 4)
        # Make it pass through (r1, y1)
        k = y1 - a_quad * (r1 - h)^2
        f_quad = expand(a_quad * (x - h)^2 + k)
        
        integrand = expand(f_cubic - f_quad)
        antideriv = integrate(integrand, x)
        area_val = substitute(antideriv, x => r2) - substitute(antideriv, x => r1)
        area_val = abs(simplify(area_val))
        
        problem(
            question="Find the area between \$f(x) = $(tex(f_cubic))\$ and \$g(x) = $(tex(f_quad))\$ between their intersection points.",
            answer=area_val,
            difficulty=(1500, 1700),
            solution=steps(
                "Solve for intersection points by setting \$f(x) = g(x)\$",
                sol("Intersection points", "x = $(r1), x = $(r2)"),
                "Determine which function is on top (check midpoint)",
                sol("Set up integral", "A = \\int_{$(r1)}^{$(r2)} |f(x) - g(x)| dx"),
                sol("Integrand", integrand),
                "Evaluate the definite integral",
                sol("Area", area_val)
            ),
            time=270
        )
        
    else  # polynomial_shifted
        # Simple polynomial minus a shifted version
        r1 = 0
        r2 = randint(3, 6)
        
        a = nonzero(-3, 3)
        b = randint(-8, 8)
        c = randint(-10, 10)
        
        f1 = a * x^2 + b * x + c
        shift = randint(5, 20)
        f2 = expand(f1 - shift)
        
        integrand = simplify(f1 - f2)
        antideriv = integrate(integrand, x)
        area_val = substitute(antideriv, x => r2) - substitute(antideriv, x => r1)
        area_val = abs(area_val)
        
        problem(
            question="Find the area between \$y = $(tex(f1))\$ and \$y = $(tex(f2))\$ from \$x = $(r1)\$ to \$x = $(r2)\$.",
            answer=area_val,
            difficulty=(1200, 1400),
            solution=steps(
                sol("Upper function", f1),
                sol("Lower function", f2),
                sol("Difference", integrand),
                sol("Set up integral", "A = \\int_{$(r1)}^{$(r2)} $(tex(integrand)) dx"),
                "Evaluate the integral",
                sol("Area", area_val)
            ),
            time=150
        )
    end
end